add vehicleRentalStationsByBbox query to gtfs graqphql schema

JustCris654 opened the pull request:
### Summary

Add vehicleRentalStationsByBbox query to gtfs graphql schema

### Unit tests

Write a few words on how the new code is tested.

- Was any manual verification done? I tested the graqhql query with the graphiql frontend
- Do all tests
  pass [the continuous integration service](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/doc/user/Developers-Guide.md#continuous-integration)
  ?

### Documentation

I added the proper comments in the schema.graphqls file

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1731,6 +1731,17 @@ type QueryType {
+  "Get all vehicle rental stations within the specified bounding box"
+  vehicleRentalStationsByBbox(
+    "Northern bound of the bounding box"
+    maxLat: Float!,
```
maximumLatitude

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1731,6 +1731,17 @@ type QueryType {
+  "Get all vehicle rental stations within the specified bounding box"
+  vehicleRentalStationsByBbox(
+    "Northern bound of the bounding box"
+    maximumLatitude: Float!,
+    "Eastern bound of the bounding box"
+    maximumLongitude: Float!,
+    "Southern bound of the bounding box"
+    minimumLatitude: Float!,
+    "Western bound of the bounding box"
+    minimumLongitude: Float!
```
Use the CoordinateValue scalar instead of Float here.

optionsome reviewed (changes requested):
You could try adding tests for this query. You can mimic what is done in for example https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/application/src/test/resources/org/opentripplanner/apis/gtfs/queries/vehicle-rental-station.graphql https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/application/src/test/resources/org/opentripplanner/apis/gtfs/expectations/vehicle-rental-station.json. The tests are run in https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/application/src/test/java/org/opentripplanner/apis/gtfs/GraphQLIntegrationTest.java

optionsome commented:
What is your use case for this btw? Are you only interested in the stations, not floating vehicles? Would it make more sense to use a vector map layer to achieve something similar to this?

JustCris654 commented:
If possible I prefer to keep it a graphql query because we have a backend that calls OTP and add some custom data to the returned object.
You're right about the floating vehicles, I need also them, what's the best way to add them? I should add another query for them because the objects are different for stations and floating vehicles, right?

optionsome commented:
> You're right about the floating vehicles, I need also them, what's the best way to add them? I should add another query for them because the objects are different for stations and floating vehicles, right?

There are two options. Either separate queries or one query that returns a new interface/union type which we could call `VehicleRentalPlace`. Perhaps using an interface could make sense here since these two types have some fields in common. Then you can call the query something like `vehicleRentalPlacesByBbox` return `[VehicleRentalPlace!]!` (for lists, I think we should always aim to make the list and its values non-null, return empty list if nothing is found). I think we call the interface VehicleRentalPlace within OTP code but we can also discuss if that is the best name in this context and do we need to prefix or postfix it with the word "Interface".

JustCris654 commented:
I think the easiest way to do this is to have two different queries for stations and free floating but I also like the idea of having one single query that does both.
I thought of something like this but I'm not sure of the implementation because I surely can't use the `VehicleRentalPlace` interface to represent every field, for example `availableVehicles: RentalVehicleEntityCounts` can't be added in `VehicleRentalPlace` in my opinion because logically is a property of a station. We could add only the number of available vehicles but we would lose their types so it's still not the best solution.
Do you have any opinion about this? 

```graphql
type VehicleRentalPlace implements Node PlaceInterface {
  """
  If true, vehicles can be returned to this station if the station has spaces available
  or allows overloading.
  """
  allowDropoff: Boolean
  "If true, vehicles can be currently returned to this station."
  allowDropoffNow: Boolean
  "If true, vehicles can be returned even if spacesAvailable is zero or vehicles > capacity."
  allowOverloading: Boolean
  "If true, vehicles can be picked up from this station if the station has vehicles available."
  allowPickup: Boolean
  "If true, vehicles can be currently picked up from this station."
  allowPickupNow: Boolean
  "Nominal capacity (number of racks) of the rental station. One if free floating vehicle."
  capacity: Int
  "Number of vehicles currently available on the rental station, grouped by vehicle type."
  availableVehicles: RentalVehicleEntityCounts
  "Global object ID provided by Relay. This value can be used to refetch this object using **node** query."
  id: ID!
  "Latitude of the vehicle (WGS 84)"
  lat: Float
  "Longitude of the vehicle (WGS 84)"
  lon: Float
  "Name of the vehicle"
  name: String!
  "If true, station is on and in service."
  operative: Boolean
  "The vehicle rental network information. This is referred as system in the GBFS terminology."
  rentalNetwork: VehicleRentalNetwork!
  "Platform-specific URLs to begin the vehicle."
  rentalUris: VehicleRentalUris
  "ID of the vehicle in the format of network:id"
  stationId: String
  "Vehicle current fuel percentage. Highest battery percent among vehicles if it is a station."
  currentFuelPercent: Float
}
```

JustCris654 commented:
I added also the currentFuelPercent property, it is related to an [issue](https://github.com/opentripplanner/OpenTripPlanner/issues/6192) I opened a week ago. Of course I will not implement it here.

optionsome commented:
> I thought of something like this but I'm not sure of the implementation because I surely can't use the `VehicleRentalPlace` interface to represent every field, for example `availableVehicles: RentalVehicleEntityCounts` can't be added in `VehicleRentalPlace` in my opinion because logically is a property of a station. We could add only the number of available vehicles but we would lose their types so it's still not the best solution. Do you have any opinion about this?

If you create a new interface, you don't have to specify all the fields from both types there, instead you can only specify the common types. So something like
```
interface VehicleRental {
  "Latitude of the vehicle rental object (WGS 84)"
  lat: Float
  "Longitude of the vehicle rental object (WGS 84)"
  lon: Float
  "If true, a vehicle can be picked up."
  allowPickupNow: Boolean
  "If true, rental object is not disabled."
  operative: Boolean
  "The vehicle rental network information. This is referred as system in the GBFS terminology."
  rentalNetwork: VehicleRentalNetwork!
}
```
Then you can change the existing types to implement this interface (which they sort of do already)
```
type VehicleRentalStation implements Node & PlaceInterface & VehicleRental

type RentalVehicle implements Node & PlaceInterface & VehicleRental
```
Then you can from `vehicleRentalPlacesByBbox` `[VehicleRentalPlace!]!`.

In practice, it means you can either use named or inlined fragments to use these types in query (this example is with inlined fragments)
```
{
  vehicleRentalPlacesByBbox(<args>) {
    lat
    lon
    ... on VehicleRentalStation {
      stationId
    }
    ... on RentalVehicle {
      vehicleId
    }
  }
}

```

JustCris654 commented:
Hi, thanks for the input. I implemented the VehicleRental interface with the relative query. 
Also added the graphql integration test and removed the vehicleRentalStationsByBbox query.

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -70,6 +70,23 @@ interface PlaceInterface {
   lon: Float
 }
 
+interface VehicleRental {
```
Add description for this interface

leonardehrenfried commented on application/src/main/java/org/opentripplanner/apis/gtfs/datafetchers/QueryTypeImpl.java:
```diff
@@ -928,6 +928,34 @@ public DataFetcher<Iterable<VehicleRentalStation>> vehicleRentalStations() {
   }
 
+  @Override
+  public DataFetcher<Iterable<Object>> vehicleRentalsByBbox() {
```
Why not return `VehicleRentalPlace` here. You probably need to change the codegen configuration: https://github.com/opentripplanner/OpenTripPlanner/blob/f4bfedfc41d3b92df45927a72edcab28624510d6/application/src/main/java/org/opentripplanner/apis/gtfs/generated/graphql-codegen.yml

JustCris654 replied:
I thought about it but couldn't get it right, I didn't know about the codegen configuration

leonardehrenfried commented on application/src/main/java/org/opentripplanner/apis/gtfs/datafetchers/QueryTypeImpl.java:
```diff
@@ -928,6 +928,34 @@ public DataFetcher<Iterable<VehicleRentalStation>> vehicleRentalStations() {
+        new Coordinate(args.getGraphQLMaximumLongitude(), args.getGraphQLMaximumLatitude())
+      );
+
+      Stream<Object> vehicleRental = Stream
+        .concat(
+          vehicleRentalService.getVehicleRentalStations().stream(),
+          vehicleRentalService.getVehicleRentalVehicles().stream()
+        )
```
```suggestion
      vehicelRentalService.getVehicleRentalPlaces()
```

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -70,6 +70,24 @@ interface PlaceInterface {
+  lat: Float
+  "Longitude of the vehicle rental object (WGS 84)"
+  lon: Float
+  "Name of the vehicle"
```
```suggestion
  "Name of the vehicle or station."
```

optionsome replied:
Actually, I'm not sure if the name should be part of this interface, in the context of rental vehicles, name is slightly weird.

leonardehrenfried replied:
I think you're right. Our internal model has a name for a free-floating vehicle but that is probably a mistake.

GBFS doesn't have it in free_bike_status: https://github.com/MobilityData/gbfs/blob/v2.3/gbfs.md#free_bike_statusjson

JustCris654 replied:
You're right! 
I'll remove the `name` property from the RentalVehicle interface then and I'll keep it only in the `VehicleRentalStation` type. 
I think the confusion arises from the `VehicleRentalPlace` interface where there are some get methods that belongs only to the `VehicleRentalStation` class implementation and not the `VehicleRentalVehicle` (free-floating vehicle). 
Some of those are:
- getCapacity: it's always 1 in the free-floating vehicle implementation
- getVehiclesAvailable: also 1
- allowDropoffNow: always false in a free-floating
Maybe I can refactor this interface in a separate PR

JustCris654 replied:
While trying to remove it from `VehicleRentalPlace.java` I noticed that the *getName* method is being used in multiple other places (like [here](https://github.com/JustCris654/OpenTripPlanner/blob/ccc614041f838cb1f23d3c133f1fb9830b3f5e6c/application/src/main/java/org/opentripplanner/service/vehiclerental/street/VehicleRentalPlaceVertex.java#L24)), I don't think I have knowledge to remove it from all this places.
What's the best way to proceed? I could remove it from the gtfs graphl API but leave it in the `VehicleRentalPlace` interface.

optionsome replied:
Unfortunately, we have to live with the existence of the rental vehicle name for backwards compatibility reasons. Therefore, in the context of this pr, it's enough that you just remove the name from the new GraphQL interface, you shouldn't remove it from the RentalVehicle type or from elsewhere.

leonardehrenfried commented on application/src/main/java/org/opentripplanner/service/vehiclerental/VehicleRentalService.java:
```diff
@@ -38,4 +38,14 @@ List<VehicleRentalStation> getVehicleRentalStationForEnvelope(
+   * Gets all vehicle rental places inside an envelope.
+   */
+  List<VehicleRentalPlace> getVehicleRentalPlacesForEnvelope(
+    double minLon,
+    double minLat,
+    double maxLon,
+    double maxLat
```
Since the method mentions the envelope can you make the it use the `Envelope` class? I know that the other method also uses the doubles but we don't want to continue this.

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -70,6 +70,22 @@ interface PlaceInterface {
 }
 
+"Interface for rental vehicles, e.g. free floating vehicles and rental vehicle stations"
+interface VehicleRental {
```
Create union for stations and free floating vehicles instead of interface.

JustCris654 replied:
I chose `RentalPlace` as the union name because I think it's effective. But if you have a better idea I'll change it

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1731,6 +1734,17 @@ type QueryType {
+    minimumLatitude: CoordinateValue!,
+    "Western bound of the bounding box"
+    minimumLongitude: CoordinateValue!
+  ): [RentalPlace]!
```
```suggestion
  ): [RentalPlace!]!
```
